go/vt/vtgate/scatter_conn.go

```diff
@@ -185,23 +188,25 @@ func (stc *ScatterConn) ExecuteMultiShard(
 		session,
 		notInTransaction,
 		func(target *querypb.Target, shouldBegin bool, transactionID int64) (int64, error) {
-			var innerqr *sqltypes.Result
-			var opts *querypb.ExecuteOptions
+			var (
+				innerqr *sqltypes.Result
+				err     error
+				opts    *querypb.ExecuteOptions
+			)
 			if session != nil && session.Session != nil {
 				opts = session.Session.Options
 			}
-			if shouldBegin {
-				var err error
+
+			switch {
+			case autocommit:
```

Comment:
The tight interplay between the autocommit state and the various layers of `notInTransaction` and `shouldBegin` causes me a bit of concern here.

Would it maybe make more sense to move the autocommit check inside `multiGoTransaction` and then change the callback to handle the three possibilities more explicitly?

Reply:
multiGoTransaction is shared with V2 code. So, we can't touch it. Which reminds me: I should add a comment that this is a V3 only function.

Fortunately, `notInTransaction` is a V2-only thing. We can assume that it's always true for V3.

So, the only interaction is with `shouldBegin`. In my initial iteration, I had the autocommit check inside the `shouldBegin` block, implying that autocommit can be relevant only if shouldBegin is true.

The downside of that approach: we obtain the 'token' to commit from the session. But then, what if shouldBegin is somehow false? In that case we don't commit anything. So, I thought it's better to make it obvious that autocommit will always happen if the flag is set.

I'm wondering if we should make a V3 version of multiGoTransaction instead.

Reply:
Personally I think it would be great if the v3 code paths were a bit more isolated from v2 since we will occasionally continue to run into these issues where it's unclear what support is needed for which mode.

Reply:
I thought about this some more. The multiGoTransaction function was a little too big to duplicate. So, I think it's worth reusing for now. Longer term, we'll get to clean this up when we deprecate v2.